### ISSUE TITLE
Consultation

### ISSUE THREAD

**Issue by cong-code** [JohnLaMaster/Impartial-Multi-Task-Learning#1]
Hello! I would like to ask about the sixth one: Nothing needs to be done to shared or specific as their .backward() calls are used in self.IMTL(), Why? return torch.sum(L_t * alpha),  Why doesn't he need to update the backward()?

**Comment by JohnLaMaster** [comment]
Hi @cong-code. I just saw your message. I feel like it's been a VERY long time since I developed this. But here's what I remember and what I've read delving into this today. The gradients of _L_t_ are calculated when defining _g_t_. _g_t_ is then used to define _alpha_. The purpose of the _.backward()_ call is to calculate the gradients, but that was already done for _alpha(g_t)_ on line 42:
```
g_t[i,::] = torch.autograd.grad(L_t[i], data, retain_graph=True, create_graph=True)[0].mean(dim=0, keepdim=True)
```
Because the gradients were already calculated when defining _alpha_, you don't need another backward(). Therefore, you can go straight to:
```
self.optimizer.step()
self.optimizer.zero_grad()
```
This is only the case for values calculated using _g_t_, i.e. the gradient balance code. To incorporate the effect of balancing the losses, those gradients must still be calculated which is what you see in the return statements.

With PyTorch models, gradients can be calculated separately for any variety of loss functions. When the losses are summed **prior** to calculating the gradients with _.backward()_, the gradients become a function of all of the summed loss functions. However, multiple sets of gradients can be calculated independently and nothing NEEDS to be done with them after calculating them. When the optimizer.step() function is called, every single gradient for the parameters included in the optimizer get updated all at once. This is why the return statements do not call backward() with ```torch.sum(L_t * alpha)``` but do with the elements of ```L_t```.

Does this answer your question?

**Comment by ZZX-11** [comment]
> Because the gradients were already calculated when defining alpha, you don't need another backward() Therefore, you can go straight to:  `self.optimizer.step()  self.optimizer.zero_grad()`

I think `optimizer.step() ` will not automated calculate the `g_t[i,::]`  for the shared parameters, and you should set `data.grad=shared_loss` then optimizer can update it
